### ISSUE TITLE
Getting pd_table indices

### ISSUE THREAD

**Issue by EricMibu** [pgmpy/pgmpy#1025]
### Subject of the issue
I have created a Bayesian model for my data, and I have added the cpds for each node to the model. Now, for each record, I would like to extract the likelihood of a variable’s value given the values of its parent nodes.

If I knew the indices (say i, j, and k) for my variable values in the CPD Table of that variable (say the cpd table has 3 indices, one for the variable itself and two parent nodes), I could do an inference by calling something like this:

```
est = BayesianEstimator(my_model_name, my_dat_frame)
cpd_table =  est.estimate_cpd('variable_name', prior_type='BDeu', equivalent_sample_size=10)
likelihood_value = cpd_table.values[i][j][k]
```

Is there any code that does something like this?

Alternatively, is there way to modify the VariableElimination( model_name).query  method so that the evidence parameter can be dynamically specified based on the different number of parents for each node?

### Your environment
* pgmpy version
* Python version
* Operating System

### Steps to reproduce
Tell us how to reproduce this issue. Please provide the complete code if possible.

### Expected behaviour
Tell us what should happen

### Actual behaviour
Tell us what happens instead

**Comment by ankurankan** [comment]
Internally pgmpy implements `TabularCPD` as a subclass of `DiscreteFactor`. So, each `TabularCPD` class should have  `variables` (in case of `TabularCPD` variables = variable + evidence) and `values` attributes. The `values` is a multidimensional numpy array with each axis representing a variable in `variables`. And the variable axis is in the same order as in `variables`, so the first variable in `variables` will be represented on axis 0 of `values`, second variable on axis 1 and so on. Therefore to access any specific index you can check the order of variables in `variables` and then do a simple numpy indexing on `values`.

**Comment by EricMibu** [comment]
Hey Ankur and team,
 
Thanks a lot for your response on this. Please allow me to clarify my question better.
 
Imagine my model has three dependent variables as follows: weight, height and bmi, and my graph at node bmi looks as follows: (weight -> bmi), (height -> bmi). That is, the bmi variable has two parents in the bayesian model: height and weight. That is, the distribution is P(bmi | weight, height).
 
Imagine that BMI can take on three different value: b, d, f; weight can take on three different values as follows: l, m, h; and height can take on three different values: t, s, a. 
 
When I use the estimate_cpd command to fetch the cpd table for the bmi node, it might look something like this.
[[0.34,0.11,0.35],[0.37,0.21,0.74],[0.38,0.21,0.43]], 
[[0.11,0.21,0.33],[0.64,0.41,0.34],[0.44,0.11,0.54]], 
[[0.64,0.01,0.40],[0.34,0.21,0.34],[0.17,0.22,0.44]
]
As you have explained, I understand that the axes correspond to the order of variables. So if my cpd_table.variables method returns ['bmi', 'weight', 'height'], then it means that the first axis corresponds to the bmi variable, second axis corresponds to the weight variable and the third axis corresponds to the height variable.
 
Now, as far as I understand the cpd table does not contain the values of the three different variables. Rather it contains the probability distributions - these are the example floats shown in the 3-d array above. 
 
My question, therefore, is, taking the first axis, for example, which corresponds to the bmi variable, how do I know which of the indices 0, 1 or 2 corresponds to b, or d or f (the three different values that bmi can take)?
 
Likewise, when I go to the second axis corresponding to weight, how do I know which of the indices 0, 1 or 2 corresponds to l, or m or h (the three different values that weight can take)?
 
Thanks again for all your help on this.
 
Eric.
 
----- Original message -----From: Ankur Ankan <notifications@github.com>To: pgmpy/pgmpy <pgmpy@noreply.github.com>Cc: EricMibu <ericmibu@ke.ibm.com>, Author <author@noreply.github.com>Subject: Re: [pgmpy/pgmpy] Getting pd_table indices (#1025)Date: Tue, Sep 11, 2018 10:26 AM 
Internally pgmpy implements TabularCPD as a subclass of DiscreteFactor. So, each TabularCPD class should have variables (in case of TabularCPD variables = variable + evidence) and values attributes. The values is a multidimensional numpy array with each axis representing a variable in variables. And the variable axis is in the same order as in variables, so the first variable in variables will be represented on axis 0 of values, second variable on axis 1 and so on. Therefore to access any specific index you can check the order of variables in variables and then do a simple numpy indexing on values.

**Comment by Kwanikaze** [comment]
I'm using the 'assignment' method which returns a list of assignments for the corresponding index.

**Comment by ankurankan** [comment]
@EricMibu Sorry about the super late reply. For checking the name-index assignments you can have a look at the `get_state_names` and `get_state_no` methods of `TabularCPD`. The `get_state_names` would return the name of the state corresponding to the index.

**Comment by Kwanikaze** [comment]
I receive an error:
AttributeError: 'TabularCPD' object has no attribute 'get_state_names'

**Comment by ankurankan** [comment]
@Kwanikaze That shouldn't happen. Which version of pgmpy (`pgmpy.__version__`) are you using? You can also try updating to the latest dev branch.

**Comment by Kwanikaze** [comment]
Hi, the latest version using Conda (pgmpy                     0.1.7                    py37_0    ankurankan)
The object is of type: <class 'pgmpy.factors.discrete.CPD.TabularCPD'>

Edit: 
Tried converting the object to a DiscreteFactor so it's type: <class 'pgmpy.factors.discrete.DiscreteFactor.DiscreteFactor'>

But still receive an error:
AttributeError: 'DiscreteFactor' object has no attribute 'get_state_no'

Note: I used BayesianModel.fit to obtain a list of TabularCPD objects.

Second Edit:
With a TabularCPD object, '.state_names' works but how do I get all the corresponding 'state_no's? I suppose it isn't hard to compute myself since the indexes of the state names would correspond to the state_no.
